### ISSUE TITLE
Items that can be spared from the state

### ISSUE THREAD

**Issue by kwalcock** [lum-ai/odinson#192]
If I find a ResultItem

```scala
case class ResultItem(
  segmentDocBase: Int, segmentDocId: Int, docIndex: Int,
  label: String, name: String,
  odinsonMatch: OdinsonMatch
)
```
which has no name or label, do I need to put it into the state?  If it has neither, but the odinsonMatch has a namedCapture
```scala
case class NamedCapture(name: String, label: Option[String], capturedMatch: OdinsonMatch)
```
that does have a name or label, do I need to put the whole tree in?  While running the tests I see quite a few things going into the State which might not be retrievable and it could save time to ignore them.

**Comment by BeckySharp** [comment]
These are great questions that we should discuss.
> If I find a ResultItem ... which has no name or label, do I need to put it into the state? 

If there is no label or name, I'd say no.  We have no way to retrieve them.

Re: 
> If it has neither, but the odinsonMatch has a namedCapture ... that does have a name or label, do I need to put the whole tree in?

I think you'd only put ONLY the named capture into the state, since the outer structure doesn't have a name etc.  

@marcovzla thoughts?

**Comment by marcovzla** [comment]
I don't think we can have a top level mention without a label, because the label has to be specified in the rule.
For example:
```
  - name: rule_1
    label: Person
    pattern: |
      Keith
```
then the mention has to be labeled according to the label provided in the rule.

The same is true for event mentions:
```
  - name: rule_2
    label: Code
    pattern: |
      trigger = [lemma=code]
      programmer: Person = >nsubj
```
the event must have the label `Code`.

There is another entrypoint in `ExtractorEngine` for executing an arbitrary pattern (not a rule).
I believe those shouldn't go into the state, because they are meant to be used interactively and the results should be displayed in a UI. If for some reason the user likes the results and wants to store them in a state then that could be a second step and the user should specify a label at that point.



However, I can think of one exception. We support the following rule syntax:
```
  - name: rule_3
    label: Code
    pattern: |
      trigger = [lemma=code]
      programmer = >nsubj [tag=NNP]
```
which would create a mention on the fly for the programmer argument, and that mention wouldn't have a label (note the argument doesn't have a label like the one above).

I think that mention should go into the state. It is true that we can't retrieve it directly because it has no label, however, when we retrieve the event mention that contains it, the unlabeled mention should still be included as an argument.

**Comment by kwalcock** [comment]
If the event mention that contains it gets sent to the state, I do include all its children so that it can be reconstituted correctly.  However, I also look at the children separately to see if they need to be added to the state as separate top-level values that can be searched for.  I don't know if that is necessary.  BTW for those without a label, it is entered as "", so that could be used to find them.

**Comment by BeckySharp** [comment]
is it only necessary to add the children when they have `promote == true`?
